### PR TITLE
Add overview of geoscience stack

### PR DESCRIPTION
**Pull request by DamienIrving**
I've had a go at expanding the [overview](https://foundations.projectpythia.org/core/overview.html) of the Python geoscience software stack.

The content I've added has been adapted from [this page](https://pyaos.github.io/stack/) on the PyAOS website. Incorporating some of the PyAOS content into Pythia Foundations is something we've been [discussing](https://github.com/ProjectPythia/projectpythia.github.io/discussions/210) for a while.

### CODE REVIEW

File: core/overview.md [hunk 1]
```diff
@@ -1,10 +1,118 @@
...
+## Core libraries
 
-## Packages
+Most geoscience data analysis involves working with numerical arrays.
+The default library for dealing with numerical arrays in Python is [NumPy](http://www.numpy.org/).
+It has some built in functions for calculating very simple statistics
+(e.g. maximum, mean, standard deviation),
+but for more complex analysis
+(e.g. interpolation, integration, linear algebra)
+the [SciPy](https://www.scipy.org/scipylib/index.html) library is the default.
```
Comment (brian-rose) [inline core/overview.md:16]:
```suggestion
the [SciPy](https://scipy.org) library is the default.
```

To fix the broken link!

File: core/overview.md [hunk 2]
```diff
@@ -1,10 +1,118 @@
...
+
+Another common feature of geo-data science is time series analysis.
+The Python standard library comes with a [datetime](https://docs.python.org/3/library/datetime.html)
+package for manipulating dates and times.
+NumPy also includes a `datetime64` module for efficient vectorized datetime operations
+and the [cftime](https://unidata.github.io/cftime/) library
+is useful for dealing with non-standard calendars.
+
+When it comes to data visualization,
+the default library is [matplotlib](https://matplotlib.org/).
```
Comment (brian-rose) [inline core/overview.md:30]:
```suggestion
the default library is [Matplotlib](https://matplotlib.org/).
```

File: core/overview.md [hunk 3]
```diff
@@ -1,10 +1,118 @@
...
+As you can see at the [matplotlib gallery](https://matplotlib.org/stable/gallery/index.html),
```
Comment (brian-rose) [inline core/overview.md:31]:
```suggestion
As you can see at the [Matplotlib gallery](https://matplotlib.org/stable/gallery/index.html),
```

File: core/overview.md [hunk 4]
```diff
@@ -1,10 +1,118 @@
...
+this library is great for any simple (e.g. bar charts, contour plots, line graphs),
+static (e.g. .png, .eps, .pdf) plots.
+The [cartopy](https://scitools.org.uk/cartopy/docs/latest/) library
```
Comment (brian-rose) [inline core/overview.md:34]:
```suggestion
The [Cartopy](https://scitools.org.uk/cartopy/docs/latest/) library
```

File: core/overview.md [hunk 5]
```diff
@@ -1,10 +1,118 @@
...
+could be achieved with a combination of the core libraries,
+their flexible, all-purpose nature means relatively common/simple tasks
+can often require quite a bit of work (i.e. many lines of code).
+To make things more efficient for data scientists,
+the scientific Python community has therefore built a number of libraries on top of the core stack.
+These high-levels libraries aren’t as flexible
+– they can’t do _everything_ like the core stack can –
+but they can do common tasks with far less effort.
+
+The most popular high-level data science library is undoubtedly [pandas](http://pandas.pydata.org/).
```
Comment (brian-rose) [inline core/overview.md:49]:
```suggestion
The most popular high-level data science library is undoubtedly [Pandas](http://pandas.pydata.org/).
```

File: core/overview.md [hunk 6]
```diff
@@ -1,10 +1,118 @@
...
+The key advance offered by pandas is the concept of labelled arrays.
```
Comment (brian-rose) [inline core/overview.md:50]:
```suggestion
The key advance offered by Pandas is the concept of labelled arrays.
```

File: core/overview.md [hunk 7]
```diff
@@ -1,10 +1,118 @@
...
+Rather than referring to the individual elements of a data array using a numeric index
+(as is required with NumPy),
+the actual row and column headings can be used.
+That means information from the cardiac ward on 3 July 2005
+could be obtained from a medical dataset by asking for `data['cardiac'].loc['2005-07-03']`,
+rather than having to remember the numeric index corresponding to that ward and date.
+This labelled array feature,
+combined with a bunch of other features that streamline common statistical and plotting tasks
+traditionally performed with SciPy, datetime and matplotlib,
```
Comment (brian-rose) [inline core/overview.md:59]:
```suggestion
traditionally performed with SciPy, datetime and Matplotlib,
```

File: core/overview.md [hunk 8]
```diff
@@ -1,10 +1,118 @@
...
+greatly simplifies the code development process (read: less lines of code).
+
+One of the limitations of pandas
```
Comment (brian-rose) [inline core/overview.md:62]:
```suggestion
One of the limitations of Pandas
```

File: core/overview.md [hunk 9]
```diff
@@ -1,10 +1,118 @@
...
+is that it’s only able to handle one- or two-dimensional (i.e. tabular) data arrays.
+The [xarray](http://xarray.pydata.org/) library was therefore created
```
Comment (brian-rose) [inline core/overview.md:64]:
```suggestion
The [Xarray](http://xarray.pydata.org/) library was therefore created
```

File: core/overview.md [hunk 10]
```diff
@@ -1,10 +1,118 @@
...
+to extend the labelled array concept to x-dimensional arrays.
+Not all of the pandas functionality is available
```
Comment (brian-rose) [inline core/overview.md:66]:
```suggestion
Not all of the Pandas functionality is available
```

File: core/overview.md [hunk 11]
```diff
@@ -1,10 +1,118 @@
...
+(which is a trade-off associated with being able to handle multi-dimensional arrays),
+but the ability to refer to array elements by their actual latitude (e.g. 20 South),
+longitude (e.g. 50 East), height (e.g. 500 hPa) and time (e.g. 2015-04-27), for example,
+makes the xarray data array far easier to deal with than the NumPy array.
```
Comment (brian-rose) [inline core/overview.md:70]:
```suggestion
makes the Xarray data array far easier to deal with than the NumPy array.
```

File: core/overview.md [hunk 12]
```diff
@@ -1,10 +1,118 @@
...
+As an added bonus,
+xarray also has built in functionality for reading/writing specific geoscience file formats
```
Comment (brian-rose) [inline core/overview.md:72]:
```suggestion
Xarray also has built in functionality for reading/writing specific geoscience file formats
```

File: core/overview.md [hunk 13]
```diff
@@ -1,10 +1,118 @@
...
+(e.g netCDF, GRIB)
+and incorporates Dask under the hood to make dealing with large arrays easier.
+
+You will occasionally find yourself needing to use a core library directly
+(e.g. you might create a plot with xarray and then call a specific matplotlib
```
Comment (brian-rose) [inline core/overview.md:77]:
```suggestion
(e.g. you might create a plot with Xarray and then call a specific Matplotlib
```

File: core/overview.md [hunk 14]
```diff
@@ -1,10 +1,118 @@
...
+function to customise a label on that plot),
+but to avoid re-inventing the wheel your first impulse should always be
+to check whether a high-level library like pandas or xarray has the functionality you need.
```
Comment (brian-rose) [inline core/overview.md:80]:
```suggestion
to check whether a high-level library like Pandas or Xarray has the functionality you need.
```

File: core/overview.md [hunk 15]
```diff
@@ -1,10 +1,118 @@
...
+Nothing would be more heartbreaking than spending hours writing your own function
+using the netCDF4 library for extracting the metadata contained within a netCDF file,
+for instance,
+only to find that xarray automatically keeps this information upon reading a netCDF file.
```
Comment (brian-rose) [inline core/overview.md:84]:
```suggestion
only to find that Xarray automatically keeps this information upon reading a netCDF file.
```

File: core/overview.md [hunk 16]
```diff
@@ -1,10 +1,118 @@
...
+In this way, a solid working knowledge of the geoscience stack
+can save you a lot of time and effort.
+
+## Domain-specific libraries
+
+So far we’ve considered libraries that do general,
+broad-scale tasks like data input/output, common statistics, visualisation, etc.
+Given their large user base,
+these libraries are usually written and supported by large companies/institutions
+(e.g. the MetOffice supports cartopy)
```
Comment (brian-rose) [inline core/overview.md:94]:
```suggestion
(e.g. the MetOffice supports Cartopy)
```

File: core/overview.md [hunk 17]
```diff
@@ -1,10 +1,118 @@
...
+or the wider PyData community (e.g. NumPy, pandas, xarray).
```
Comment (brian-rose) [inline core/overview.md:95]:
```suggestion
or the wider PyData community (e.g. NumPy, Pandas, Xarray).
```

File: core/overview.md [hunk 18]
```diff
@@ -1,10 +1,118 @@
...
+Within each sub-discipline of the geosciences,
+individuals and research groups take these general libraries
+and apply them to their very specific data analysis tasks.
+Increasingly, these individuals and groups
+are formally packaging and releasing their code for use within their community.
+For instance, Andrew Dawson (an atmospheric scientist at Oxford)
+does a lot of EOF analysis and manipulation of wind data,
+so he has released his [eofs](https://ajdawson.github.io/eofs/latest/)
+and [windspharm](https://ajdawson.github.io/windspharm/latest/) libraries
+(which are able to handle data arrays from NumPy or xarray).
```
Comment (brian-rose) [inline core/overview.md:105]:
```suggestion
(which are able to handle data arrays from NumPy or Xarray).
```

**Review by brian-rose — COMMENTED** [review]
Hi @DamienIrving, thanks for getting this started!

I just went through quickly and added several suggested edits to use consistent capitalization of library names.

I think this is a really nice starting point for our presentation of the stack.

File: core/overview.md [hunk 19]
```diff
@@ -1,10 +1,118 @@
...
+the [SciPy](https://www.scipy.org/scipylib/index.html) library is the default.
+If you’re dealing with particularly large arrays,
+[Dask](https://dask.org/) works with the existing Python ecosystem
+(including NumPy) to scale your analysis
+to multi-core machines and/or distributed clusters (i.e. parallel processing).
+
+Another common feature of geo-data science is time series analysis.
+The Python standard library comes with a [datetime](https://docs.python.org/3/library/datetime.html)
+package for manipulating dates and times.
+NumPy also includes a `datetime64` module for efficient vectorized datetime operations
```
Comment (mgrover1) [inline core/overview.md:25]: Can you add a link to the date time module in numpy?

File: core/overview.md (hunk 6 above)
Comment (mgrover1) [reply core/overview.md:50]: Also, labelled --> labeled

**Review by mgrover1 — CHANGES_REQUESTED** [review]
Thanks for taking the time to put together this fantastic overview @DamienIrving - just a couple of minor comments. This will be a great addition to the Foundations content 😁

File: core/overview.md (hunk 19 above)
Comment (DamienIrving) [reply core/overview.md]: Addressed in https://github.com/ProjectPythia/pythia-foundations/pull/265/commits/738a3e0d816d3a0a65e564cb3f14193007c91284

File: core/overview.md (hunk 6 above)
Comment (DamienIrving) [reply core/overview.md]: > Also, labelled --> labeled

Addressed in https://github.com/ProjectPythia/pythia-foundations/pull/265/commits/738a3e0d816d3a0a65e564cb3f14193007c91284

**Comment by DamienIrving** [comment]
Thanks, @brian-rose and @mgrover1. I've gone ahead and made all the changes you suggested.

**Review by mgrover1 — APPROVED** [review]
Thanks for making those changes @DamienIrving !

File: core/overview.md [hunk 20]
```diff
@@ -1,10 +1,119 @@
...
+(which are able to handle data arrays from NumPy or Xarray).
+Similarly, a group at the Atmospheric Radiation Measurement (ARM) Climate Research Facility
+have released their Python ARM Radar Toolkit ([Py-ART](http://arm-doe.github.io/pyart/))
+for analysing weather radar data.
+
+There are too many domain specific libraries to mention here,
+but online resources such as the
+[Python for Atmosphere and Ocean Science (PyAOS) package index](https://pyaos.github.io/packages/)
+attempt to keep track of the domain-specific libraries in their field.
+
```
Comment (brian-rose) [inline core/overview.md:115]:
```suggestion
Also check out the [Pythia Resource Gallery](https://projectpythia.org/gallery.html) and try filtering by domain.

```

**Review by brian-rose — COMMENTED** [review]
Sorry just one more suggestion, with an explicit link to the Pythia Resource Gallery. Otherwise good to go!

**Comment by DamienIrving** [comment]
Done :-)

**Pull request merged**